FRAuthenticator/FRAuthenticator/Manager/AuthenticatorManager.swift

```diff
@@ -38,37 +43,63 @@ struct AuthenticatorManager {
     ///   - onFailure: failure callback with Error
     func createMechanismFromUri(uri: URL, onSuccess: @escaping MechanismCallback, onError: @escaping ErrorCallback) {
         
+        let uriType = uri.getURIType()
         let authType = uri.getAuthType()
-        FRALog.v("Received URI (\(uri.absoluteString)) with authType (\(authType.rawValue)), and proceeding to construct Mechanism object")
-
-        if authType == .hotp {
-            do {
-                let mechanism = try self.storeOathQRCode(uri: uri)
-                onSuccess(mechanism)
-            }
-            catch {
-                onError(error)
+        if uriType == .mfauth {
+            FRALog.v("Evaluating policies for the new Account")
+            let result = self.policyEvaluator.evaluate(uri: uri)
+            if let policy = result.nonCompliancePolicy, !result.comply {
+                onError(AccountError.failToRegisterPolicyViolation(policy.name))
+                return
             }
-        }
-        else if authType == .totp {
+            
+            FRALog.v("Received Combined MFA URI (\(uri.absoluteString)). Proceeding to construct Mechanism objects")
+            // try to register OATH mechanism, do not return object or error
             do {
                 let mechanism = try self.storeOathQRCode(uri: uri)
-                onSuccess(mechanism)
+                FRALog.v("OATH mechanism for (\(mechanism.issuer)) in Combined MFA successfully created.")
             }
             catch {
-                onError(error)
+                FRALog.v("Error creating OATH mechanism from Combined MFA URI")
             }
-        }
-        else if authType == .push {
+            // try to register push mechanism and do return object or error
             self.storePushQRcode(uri: uri, onSuccess: { (mechanism) in
                 onSuccess(mechanism)
             }, onFailure: {(error) in
                 onError(error)
             })
-        }
-        else {
-            FRALog.e("Unsupported auth type: \(String(describing: uri.host))")
-            onError(MechanismError.invalidType)
+        } else {
+            FRALog.v("Received URI (\(uri.absoluteString)) with authType (\(authType.rawValue)), and proceeding to construct Mechanism object")
+
```

Comment:
may be you can combine this logic ? both looks similar

if authType == .hotp || authType == .totp {
do {
                    let mechanism = try self.storeOathQRCode(uri: uri)
                    onSuccess(mechanism)
                }
                catch {
                    onError(error)
                }

}

Reply:
I agree. I am always in favor of improve pre-existing code